### DISCUSSION TITLE
Can't decorate a method with `@api_view`

### DISCUSSION THREAD

**Discussion by ZeyadMoustafaKamal** [encode/django-rest-framework#9427] [Potential Issue] [answered]
So I am trying to build something with django rest framework. and I decided to build the views in a class as methods and then decorate them with `@api_view`. but I am getting this error

AssertionError: The `request` argument must be an instance of `django.http.HttpRequest`, not `rest_admin.base.AdminSite`

and yes I am building a custom admin site.

here is the view
```Python
class Foo:
    @api_view()
    def index(self, request):
            return Response("Hello world")
    def get_urls(self):
        return [path(''), self.index, name='index], "rest_admin", "rest_admin"
 ```
and here is the URLs
```Python
foo = Foo()
urlpatterns = [
    path('admin', foo.get_urls())
]
```
after some investigations. I saw that it passes the request as the `self`   parameter so that the problem happens. anyway to fix this except to make the views outside the class ?? and also if I used `@staticmethod` decorator is this fine or it might cause some problems in the future ??

**Comment by browniebroke** [comment, marked as answer]
Correct, this decorator only works with function, not class methods. If you really insist, you may try your luck with Django's `method_decorator`: https://docs.djangoproject.com/en/5.0/ref/utils/#django.utils.decorators.method_decorator

Something along the lines of:

```python
class Foo:
    @method_decorator(api_view())
    def index(self, request):
            return Response("Hello world")
    def get_urls(self):
        return [path(''), self.index, name='index], "rest_admin", "rest_admin"
```

I have not tested it with `api_view`, but this worked for me in other similar cases.

**Reply by ZeyadMoustafaKamal** [reply]
OMG. thanks a lot.
